### PR TITLE
[embedded] Implicitly define __APPLE__ and __MACH__ when on -apple-none triples

### PR DESCRIPTION
**Pull request by kubamracek**
This removes the need for the stdlib and other modules to -Xcc those defines manually, and also drops this requirement from users of the -apple-none triples (see the lit test changes).

### CODE REVIEW

**Comment by kubamracek** [comment]
@swift-ci please test

File: lib/ClangImporter/ClangImporter.cpp [hunk 1]
```diff
@@ -631,6 +631,17 @@ importer::getNormalInvocationArguments(
       });
     }
 
+    // To support -apple-none triples, which are non-Darwin, we need to #define
+    // a few things that the Apple SDKs expect.
+    if (triple.getVendor() == llvm::Triple::VendorType::Apple) {
+      invocationArgStrs.insert(invocationArgStrs.end(),
+                               {"-D__APPLE__", "-D__MACH__"});
```
Comment (phausler) [inline lib/ClangImporter/ClangImporter.cpp:638]: mach should only be defined when the triple contains mach:
technically someone could write `xxx-apple-none-elf` (granted that would be rather strange...)

Comment (rauhul) [reply lib/ClangImporter/ClangImporter.cpp:638]: +1 this should be conditional object file format == MachO

File: lib/ClangImporter/ClangImporter.cpp [hunk 2]
```diff
@@ -631,6 +631,17 @@ importer::getNormalInvocationArguments(
       });
     }
 
+    // To support -apple-none triples, which are non-Darwin, we need to #define
+    // a few things that the Apple SDKs expect.
+    if (triple.getVendor() == llvm::Triple::VendorType::Apple) {
```
Comment (rauhul) [inline lib/ClangImporter/ClangImporter.cpp:636]: Shouldn't this be done by the driver and not the clang importer?

Comment (kubamracek) [reply lib/ClangImporter/ClangImporter.cpp:636]: Hm, that would be against the established pattern -- look at what the rest of this (pretty large) function is doing. It's setting up a ton of things, essentially "construct reasonable defaults for a Clang instance given the current Swift compilation", which my addition seems to match.

I'd say this addition shouldn't go to the driver, on the grounds that that would create an inconsistency. Maybe this whole function should go to the driver? (Even that I would probably argue against.)

Comment (rauhul) [reply lib/ClangImporter/ClangImporter.cpp:636]: I'll defer to you, its just a bit odd to splat out the platform knowledge in a bunch of places, cc @artemcm for thoughts (not blocking)

**Comment by kubamracek** [comment]
@swift-ci please test

File: lib/ClangImporter/ClangImporter.cpp (hunk 1 above)
Comment (kubamracek) [reply lib/ClangImporter/ClangImporter.cpp]: Fixed.

**Review by rauhul — APPROVED** [review]

File: lib/ClangImporter/ClangImporter.cpp (hunk 2 above)
Comment (compnerd) [reply lib/ClangImporter/ClangImporter.cpp:635]: I believe that @rauhul is correct here.  `__APPLE__` is a vendor definition, and should always be set by clang for `*-apple-*` triples. I _could_ see the argument for `__MACH__` being the clang importer as _technically_ it is the kernel identifier and not the object file format specifier, but it is used for the object file format as well. I think that sinking this into upstream clang is the right approach.

File: lib/ClangImporter/ClangImporter.cpp [hunk 3]
```diff
@@ -631,6 +631,17 @@ importer::getNormalInvocationArguments(
...
 
+    // To support -apple-none, -apple-none-macho, -unknown-none-wasm triples.
+    if (triple.getVendor() == llvm::Triple::VendorType::Apple) {
+      invocationArgStrs.insert(invocationArgStrs.end(), {"-D__APPLE__"});
+    }
+    if (triple.isOSBinFormatMachO()) {
+      invocationArgStrs.insert(invocationArgStrs.end(), {"-D__MACH__"});
+    }
+    if (triple.isOSBinFormatWasm()) {
+      invocationArgStrs.insert(invocationArgStrs.end(), {"-D__wasi__"});
```
Comment (compnerd) [inline lib/ClangImporter/ClangImporter.cpp:642]: The macro here should be `__wasm__`.  WASI is not related to the object file format but the environment (WASI uses WASM and provides a system interface).

File: lib/ClangImporter/ClangImporter.cpp (hunk 2 above)
Comment (kubamracek) [reply lib/ClangImporter/ClangImporter.cpp:635]: Oh I misunderstood then. I thought that @rauhul was talking about the *Swift driver* and that *it* should receive this logic instead (and pass it as -Xcc flags down to the compiler).

File: lib/ClangImporter/ClangImporter.cpp (hunk 3 above)
Comment (kubamracek) [reply lib/ClangImporter/ClangImporter.cpp:642]: Aha, so the reason why I seemingly needed to add `-D__wasi__` is that we have SwiftShims code that conditionalizes based on it, and we fail to build the embedded stdlib for `wasm-unknown-none-wasm` without this `-D__wasi__`. The concrete piece of code is here: https://github.com/apple/swift/blob/main/stdlib/public/SwiftShims/swift/shims/Visibility.h#L137

Sounds like you're saying we should fix Visibility.h instead?

**Comment by kubamracek** [comment]
@swift-ci please test

File: lib/ClangImporter/ClangImporter.cpp (hunk 3 above)
Comment (compnerd) [reply lib/ClangImporter/ClangImporter.cpp:642]: Yeah, that definitely looks like a bug.  That should be `__wasm__`.

Comment (kubamracek) [reply lib/ClangImporter/ClangImporter.cpp]: Fixed!

**Comment by kubamracek** [comment]
@swift-ci please test

**Pull request merged**